server/src/main/java/org/apache/druid/server/coordinator/TierLoadingState.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.server.coordinator;
+
+import org.apache.druid.timeline.SegmentId;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Contains segments being loaded or replicated in a tier.
+ */
+public class TierLoadingState
+{
+  private int lifetime;
+  private final ConcurrentHashMap<SegmentId, String> processingSegments = new ConcurrentHashMap<>();
```

Comment:
Can't a tier load / balance the same segment on multiple servers simultaneously with these changes?

Reply:
The code doesn't enforce it, and it is possible for the same segment to be replicated / moved on different servers within the same tier.
This datastructure would lead to a loss of that information  (which also happens to be the case with the current implementation of ReplicationThrottler)

Reply:
A simple example is a LoadRule within a tier with a replication factor of 2. Both replicas can be loaded in the same cycle but the TierLoadingState would only keep track of the second load

Reply:
Yes, as you mention, that is the current implementation and we have retained it here.
We are hoping to relax replication throttle limit further anyway. So changing this behaviour is not really required.

Incidentally, it is also the desirable implementation, being simpler to understand and reason about than the alternative. From the docs, definition of `replicationThrottleLimit`:
`The maximum number of segments that can be replicated at one time.`
Alternative would be: 
`The number of replica segments (i.e. non primary) that can be in the queue at any time`.

We could try an alternative implementation but it would have these issues:
- Counting each replica separately would hit the throttle limit sooner (i.e. with fewer _unique_ segments), thus increasing the time to full replication. We could always increase the throttle limit but we would have to keep increasing it whenever we decided to increase the replication level on _any_ datasource.
- Assigning 1st replica of each segment, then 2nd replica of each segment is not an option either because we must make all decisions for a given segment in a single invocation of the `LoadRule.run()`.

Hope that clarifies the design choice 🙂

Reply:
However, this PR does modify the definition of `replicationThrottleLimit` and `maxSegmentsInNodeLoadingQueue` slightly (docs to be updated). The limits now apply on number of assignments done during a single coordinator run rather than items in the queue at a given instant.
This change was done to:
- make the configs easier to use
- ensure that the configs actually take effect. In some cases (where loading happens quickly enough), these limits are constantly violated.